src/ring_theory/unique_factorization_domain.lean

```diff
@@ -361,10 +361,51 @@ unique_factorization_monoid.of_exists_prime_factors (by
   { convert eif,
     simp_rw irreducible_iff_prime_of_exists_unique_irreducible_factors eif uif })
 
+namespace unique_factorization_monoid
+variables [comm_cancel_monoid_with_zero α] [decidable_eq α]
+variables [unique_factorization_monoid α]
+/-- Noncomputably determines the multiset of prime factors. -/
+noncomputable def factors (a : α) : multiset α := if h : a = 0 then 0 else
+classical.some (unique_factorization_monoid.exists_prime_factors a h)
+
+theorem factors_prod {a : α} (ane0 : a ≠ 0) : associated (factors a).prod a :=
+begin
+  rw [factors, dif_neg ane0],
+  exact (classical.some_spec (exists_prime_factors a ane0)).2
+end
+
+theorem prime_of_factor {a : α} : ∀ (x : α), x ∈ factors a → prime x :=
+begin
+  rw [factors],
+  split_ifs with ane0, { simp only [multiset.not_mem_zero, forall_false_left, forall_const] },
+  intros x hx,
+  exact (classical.some_spec (unique_factorization_monoid.exists_prime_factors a ane0)).1 x hx,
+end
+
+theorem irreducible_of_factor {a : α} : ∀ (x : α), x ∈ factors a → irreducible x :=
+λ x h, (prime_of_factor x h).irreducible
+
+lemma exists_mem_factors_of_dvd {a p : α} (ha0 : a ≠ 0) (hp : irreducible p) : p ∣ a →
+  ∃ q ∈ factors a, p ~ᵤ q :=
+λ ⟨b, hb⟩,
+have hb0 : b ≠ 0, from λ hb0, by simp * at *,
+have multiset.rel associated (p ::ₘ factors b) (factors a),
+  from factors_unique
+    (λ x hx, (multiset.mem_cons.1 hx).elim (λ h, h.symm ▸ hp) (irreducible_of_factor _))
+    irreducible_of_factor
+    (associated.symm $ calc multiset.prod (factors a) ~ᵤ a : factors_prod ha0
+      ... = p * b : hb
+      ... ~ᵤ multiset.prod (p ::ₘ factors b) :
+        by rw multiset.prod_cons; exact (factors_prod hb0).symm.mul_left _),
+multiset.exists_mem_of_rel_of_mem this (by simp)
+
+end unique_factorization_monoid
+
 namespace unique_factorization_monoid
 variables [comm_cancel_monoid_with_zero α] [decidable_eq α] [nontrivial α] [normalization_monoid α]
 variables [unique_factorization_monoid α]
 
+-- TODO: this could perhaps be `multiset.map normalize $ factors' a`
```

Comment:
Seems better, why not do it?

Reply:
I expected it to be more work than it turned out to be

Reply:
Maybe remove the comment then, or turn it into a comment explaining that the alternative definition isn't as promising as it looks.